template/koa-rest-api/src/framework/validation.ts

```diff
@@ -29,13 +31,31 @@ import type { Context } from 'src/types/koa';
  * { "/advertiserId": "advertiserId is required in the URL" }
  * ```
  */
-const parseInvalidFieldsFromError = ({
-  errors,
-}: z.ZodError): Record<string, string> =>
-  Object.fromEntries(
-    errors.map((err) => [`/${err.path.join('/')}`, err.message]),
+
+const parseInvalidFieldsFromError = ({ errors }: z.ZodError) => {
+  const invalidFields: InvalidFields = {};
+  const flattenedIssues: Array<[z.ZodIssue, UnionErrorIdx]> = errors.map(
+    (issue) => [issue, undefined],
   );
 
+  for (const [issue, unionIdx] of flattenedIssues) {
+    if (issue.code === ZodIssueCode.invalid_union) {
+      issue.unionErrors.forEach((err, idx) => {
+        err.errors.forEach((childIssue) => {
+          flattenedIssues.push([childIssue, idx]);
+        });
+      });
+    } else {
+      const path = `/${issue.path.join('/')}${
+        unionIdx !== undefined ? `_union${unionIdx}` : ''
```

Comment:
Optional suggestion: while we can't really uphold JSON Pointer compatibility anymore with this approach, perhaps using the same [initial escape character](https://datatracker.ietf.org/doc/html/rfc6901#section-3) would be more immediately recognisable than an underscore, which is relatively common in JSON property names:

```suggestion
        unionIdx !== undefined ? `~union${unionIdx}` : ''
```